packages/test/test-end-to-end-tests/src/test/attachLifecycle.spec.ts

```diff
@@ -0,0 +1,148 @@
+                            await timeoutPromise<void>(
+                                (resolve)=>initContainer.once("saved", ()=>resolve()),
+                                {durationMs: this.timeout() / 2,errorMsg:"ddsSaveAfterAttach timeout"});
+                    }
+                }
+                if(testConfig.ddsAttachPoint === 2) {
+                    // point 2 - at dds create
+                    await attachDds();
+                }
+
+                for(const i of sharedPoints) {
+                    // also send an op at these points
+                    // we'll use these to validate
+                    newMap.set(i.toString(),i);
+
+                    if(testConfig.containerAttachPoint === i) {
+                        await initContainer.attach(provider.driver.createCreateNewRequest(provider.documentId));
+                        containerUrl = initContainer.resolvedUrl;
+                    }
+                    if(testConfig.datastoreAttachPoint === i) {
+                        await attachDatastore()
+                    }
+                    if(testConfig.ddsAttachPoint === i) {
+                        await attachDds();
+                    }
+                }
+                while(initContainer.isDirty){
+                    await timeoutPromise<void>(
+                        (resolve)=>initContainer.once("saved", ()=>resolve()),
+                        {durationMs: this.timeout() / 2,errorMsg:"final save timeout"});
+                }
+
+                initContainer.close();
+            }
+            {
+                const validationLoader = createLoader(
+                    [[provider.defaultCodeDetails, provider.createFluidEntryPoint()]],
+                    provider.documentServiceFactory,
+                    provider.urlResolver,
+                );
+                const validationContainer = await validationLoader.resolve({
+                    url: await provider.driver.createContainerUrl(provider.documentId, containerUrl),
+                });
+
+                const initDataObject = await requestFluidObject<ITestFluidObject>(validationContainer, "default");
+
+                const newds = await (await timeoutAwait(
+                    initDataObject.root.wait<IFluidHandle<ITestFluidObject>>("ds"),
+                    {durationMs: this.timeout() / 2,errorMsg:"Datastore not available before timeout"})
+                ).get();
+
+                const newMap = await (await timeoutAwait(
+                        newds.root.wait<IFluidHandle<ISharedMap>>("map"),
+                        {durationMs: this.timeout() / 2,errorMsg:"Map not available before timeout"})
+                    ).get();
+
+                for(const i of sharedPoints) {
+                    assert.equal(
+                        await timeoutAwait(
+                            newMap.wait<number>(i.toString()),
+                            {durationMs: this.timeout() / 2,errorMsg:"Key not available before timeout"}),
```

Comment:
`this.timeout()` must be Mocha's timeout for this test?  Might be clearer why you're halving it if you define a const with a brief comment and use that everywhere.